Update email block core/paragraph renderer [MAILPOET-5642]

lysyjan opened the pull request:
## Description

We had a simple core/paragraph block render that didn't support all possible configurations. In this renderer iteration, we add the possibility of text alignment and rendering compatibility across email clients.

## Code review notes

_N/A_

## QA notes

QA could be skipped here, I would let the decision on the reviewer.

## Linked PRs

_N/A_

## Linked tickets

[MAILPOET-5642]

## After-merge notes

_N/A_

## Tasks

- [x] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [x] I added sufficient test coverage
- [x] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes


[MAILPOET-5642]: https://mailpoet.atlassian.net/browse/MAILPOET-5642?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

lysyjan committed:
Reuse function from SettingsController in Heading renderer

[MAILPOET-5642]

lysyjan committed:
Disable dropCap settings for Paragraphs

[MAILPOET-5642]

lysyjan committed:
Disable fontWeight settings for Paragraphs

[MAILPOET-5642]

lysyjan committed:
Fix stretch alignment for the column/core block

[MAILPOET-5642]

lysyjan committed:
Implement rendering block core/paragraph

[MAILPOET-5642]

lysyjan committed:
Extend block paragraph integration test

[MAILPOET-5642]

lysyjan committed:
Remove exception in width block width calculation

[MAILPOET-5642]

lysyjan committed:
Skip width style setting in blocks paragraph and heading

[MAILPOET-5642]

lysyjan committed:
Fix blocks width across email clients

[MAILPOET-5642]

costasovo commented on mailpoet/lib/EmailEditor/Integrations/Core/Renderer/Blocks/Paragraph.php:
```diff
@@ -28,20 +44,78 @@ private function getBlockWrapper(array $parsedBlock, SettingsController $setting
+
+        $cssString = $availableStylesheets ?? '';
+
+        $colorRegex = "/--wp--preset--color--$colorKey: (#[0-9a-fA-F]{6});/";
```
Why are we adding `--wp--preset--color--` here? Is that a convention for specifying colors in block themes? Is it documented somewhere? I think it would be helpful to have a comment explaining this here. Or maybe a comment describing a bit higher level how we get the color values. 🙏

costasovo commented on mailpoet/lib/EmailEditor/Integrations/Core/Renderer/Blocks/Paragraph.php:
```diff
@@ -28,20 +44,78 @@ private function getBlockWrapper(array $parsedBlock, SettingsController $setting
+  /**
+   * @param array{tag_name: string, class_name?: string} $tag
+   */
+  private function removePaddingFromElement($blockContent, array $tag) {
```
I guess we could add a return typehit here

costasovo commented on mailpoet/lib/EmailEditor/Engine/Renderer/Preprocessors/BlocksWidthPreprocessor.php:
```diff
@@ -34,10 +34,7 @@ public function preprocess(array $parsedBlocks, array $layoutStyles): array {
       $modifiedLayoutStyles['padding']['right'] = $block['attrs']['style']['spacing']['padding']['right'] ?? '0px';
 
-      // Set current block values and reassign it to $parsedBlocks, but don't set width for blocks that are not supposed to have it
-      if (!in_array($block['blockName'], self::BLOCKS_WITHOUT_WIDTH, true)) {
```
It seems that this change made BLOCKS_WITHOUT_WIDTH constant unused. It would be better to remove it.

costasovo reviewed (changes requested):
@lysyjan Thank you for adding the support for the paragraph block!

I added a couple of comments. Just details.

Btw. I noticed that the formatting options for the paragraph in the email editor are limited compared to the formatting options in the post editor. I think it makes sense that we can't support all the options for emails. But maybe we could add support at least for `Highlight` because I find this one quite usable. What do you think? It would be an extra ticket...
![Screenshot 2024-01-03 at 16 07 14](https://github.com/mailpoet/mailpoet/assets/1082140/cb0bf201-a061-40c1-baf2-53a6c55db246)

lysyjan committed:
Unify typehint by BlockRenderer interface

[MAILPOET-5642]

lysyjan committed:
Remove unused constant

[MAILPOET-5642]

lysyjan committed:
Add comments about color variables conversion

[MAILPOET-5642]

lysyjan commented:
@costasovo Thank you for your review. I addressed all your remarks in the three new commits.

> Btw. I noticed that the formatting options for the paragraph in the email editor are limited compared to the formatting options in the post editor. I think it makes sense that we can't support all the options for emails. But maybe we could add support at least for Highlight because I find this one quite usable. What do you think? It would be an extra ticket...

I agree with you. [By the comment in the file](https://github.com/mailpoet/mailpoet/blob/4567c6d3c1a1564e6de355d634576ac40cbe7dcb/mailpoet/assets/js/src/email-editor/engine/blocks/core/rich-text.ts#L17-L18) it looks we want to add the support later. So, creating a follow-up ticket makes sense to me.

costasovo reviewed (approved):
Thanks for the changes and the comment! The comment will help me to recall why we use the prefix in the future.
